sklearn/discriminant_analysis.py

```diff
@@ -631,8 +631,10 @@ def transform(self, X):
 
         Returns
         -------
-        X_new : ndarray of shape (n_samples, n_components)
-            Transformed data.
+        X_new : ndarray of shape (n_samples, n_components) or \
+            (n_samples, min(rank, n_components))
+            Transformed data. In the case of the 'svd' solver, the shape
+            is (n_samples, min(rank, n_components)).
```

Comment:
I think this applies to both svd and eigen solvers:

```suggestion
            Transformed data. For the `'svd'` or `'eigen'` solvers,
            the shape is `(n_samples, min(rank, n_components))`, where
            `rank=scalings_.shape[0]`.
```

For reference:

https://github.com/scikit-learn/scikit-learn/blob/1d697842a36fbd93e89d7756bf899f527baf6323/sklearn/discriminant_analysis.py#L644-L647

Reply:
That makes sense. Updated suggestion:

```suggestion
            Transformed data. For the `'svd', the shape is
            `(n_samples, min(rank, n_components))`, where
            `rank=scalings_.shape[0]`.
```

Reply:
IMO, defining rank like that wouldn't really make clear what's going on in its calculation. We could instead define: 
```
`rank` is equal to the number of singular values `s` such that `s > tol`
```
I think this gets the point across better and can potentially avoid problems such as what @michen00 was running into.

Reply:
> `rank` is equal to the number of singular values `s` such that `s > tol`

We already have this defined in `tol`: (which we can extend if it is unclear)

https://github.com/scikit-learn/scikit-learn/blob/b519d6c748e64e32a86069050200a44e365e0718/sklearn/discriminant_analysis.py#L229-L233

There are two questions here:

1. How is the rank computed? (The answer is described in `tol`)
2. How do I get the rank from the estimator? (`scalings_.shape[0]`)

In context for `transform` docs where the shape is defined in the docstring, I prefer to answer "assuming we know the answer to 1, how do we get the rank?"

Reply:
I think my personal preference would be something like 
```
`rank` is equal to the number of singular values `s` 
such that `s > tol`. `rank` is accessible as `scalings_.shape[0]`
```
because if someone were to run into an issue with the output shape, I imagine the go-to reaction would be to check the documentation for `transform` section, which would read:
```
...
Transformed data. When `solver='svd'`, the shape is
`(n_samples, min(rank, n_components))`, where
`rank=scalings_.shape[0]`.
```
Then to look at the documentation for `scalings_` 
https://github.com/scikit-learn/scikit-learn/blob/ab08e4dba5f1f87b8c3395f32469a6ddb5e34f89/sklearn/discriminant_analysis.py#L278-L280

which doesn't really clarify the situation, then proceed to the documentation for `solver`:
https://github.com/scikit-learn/scikit-learn/blob/ab08e4dba5f1f87b8c3395f32469a6ddb5e34f89/sklearn/discriminant_analysis.py#L193-L201

which itself doesn't mention tolerance. So the only way to know that the output shape is affected by `tol` is to read the documentation for `tol` directly, or troll through the source code, neither of which seem that great.

I guess my personal issue could also be solved if the docs for `solver` directly referenced `svd`'s reliance on the `tol` parameter. At the end of the day, this issue seems rare at best, so it's not a big problem no matter what the final docs read.